### ISSUE TITLE
Xbox One gamepad detected as generic XInput device on Windows

### ISSUE THREAD

**Issue by austinried** [godotengine/godot#29894]
**Godot version:**
3.1.1

**OS/device including version:**
Windows 10

**Issue description:**
Xbox One controller on Windows 10 (wired) is detected as a generic XInput controller with a guid of `__XINPUT_DEVICE__` instead of anything from the SDL controller list.  This prevents me from accurately identifying the controller as something other than a generic gamepad.

Additionally, I'm confused as to why I'm getting different results from Godot on Windows vs. Godot on Linux when it comes to comparing GUIDs from the SDL Gamepad tool (available here: http://www.generalarcade.com/gamepadtool/)

### SDL Tool 1.2 Ubuntu 18.04
**Xbox One Controller**
GUID: `030000005e040000ea02000001030000`
Name: `XInput Controller`

**PS4 Controller**
GUID: `030000004c050000c405000011810000`
Name: `PS4 Controller`

### Godot 3.1.1 Ubuntu 18.04
**Xbox One Controller**
GUID: `030000005e040000ea02000001030000`
Name: `Xbox One Wireless Controller`

**PS4 Controller**
GUID: `030000004c050000c405000011810000`
Name: `Sony DualShock 4`

### SDL Tool 1.2 Windows 10
**Xbox One Controller**
GUID: `030000005e040000ff02000000007801`
Name: `XInput Controller`

**PS4 Controller**
GUID: `030000004c050000c405000000000000`
Name: `PS4 Controller`

### Godot 3.1.1 Windows 10
**Xbox One Controller**
GUID: `__XINPUT_DEVICE__`
Name: `XInput Gamepad`

**PS4 Controller**
GUID: `4c05c405000000000000504944564944`
Name: `PS4 Controller`

All tests above were conducted with the controllers wired, not wireless.

**Steps to reproduce:**
Connect a controller and use `Input.get_joy_guid(0)`.

**Comment by bruvzg** [comment]
Possibly related to #26544, Some bindings and SDL gamepad tool use unsupported 2.0.6+ binding format.

**Comment by austinried** [comment]
Actually I've just realized that I also can't find to GUID that the SDL gamepad tool shows on Windows 10 for the Xbox One controller (`030000005e040000ff02000000007801`) in either Godot's SDL GUID list or the list over here: https://github.com/gabomdq/SDL_GameControllerDB

Does the GUID change with the 2.0.6 binding format?  Or do I have a strange Xbox One controller?

**Comment by bruvzg** [comment]
> Does the GUID change with the 2.0.6 binding format?

Looks like it is changed. My Xbox One controller have 5e04000000000000d102000000000000 GUID with pre 2.0.6 version of Gamepad tool and 030000005e040000d102000000000000 with the new one.

**Comment by austinried** [comment]
I guess either way I would expect to get at least one of those back from `Input.get_joy_guid()` to even allow me to add my own mapping for it.

**Comment by piratesephiroth** [comment]
I guess there's a bunch of them.
According to generalarcade's gamepadtool, my Xbox One controller's GUID is 030000005e040000dd02000000007801, not in any of the lists either.

**Comment by salvob41** [comment]
The problem is basically described here: #8368 . We should address that issue

**Comment by akien-mga** [comment]
Can you check if it works fine in 3.2 beta 6 or later?
I might have fixed this with #34910.

**Comment by akien-mga** [comment]
I'll assume it fixed by #34910 for now, and if it isn't, it should be fixable by contributing the missing mappings to https://github.com/gabomdq/SDL_GameControllerDB.

Please comment if you can still reproduce the issue / need help to contribute mappings upstream if they're missing.

**Comment by tristongilley** [comment]
**Godot 3.2.2 Beta 3 Windows 10**

Xbox One Controller (Model 1708)
GUID: `__XINPUT_DEVICE__`

PS4 Controller (Model CUH-ZCT1U)
GUID: `4c05c405000000000000504944564944`

The PS4 is connected via USB and the Xbox I tried both wired and bluetooth.

**Steps to reproduce**:
Connect a controller and use Input.get_joy_guid(0)

**Comment by akien-mga** [comment]
Related to https://github.com/godotengine/godot/issues/38554#issuecomment-633980886

CC @madmiraal

**Comment by madmiraal** [comment]
There are two issues here (and both are specific to Windows):
1. The GUID returned by Godot, which should be fixed by #39060.
2. Godot has its own XInput device mapping, hence the [#38554 comment](https://github.com/godotengine/godot/issues/38554#issuecomment-633980886).

**Comment by akien-mga** [comment]
Can you still reproduce this issue in 3.2.3 RC 1 or the current `master` branch?

**Comment by hoontee** [comment]
@akien-mga 

3.2.3.rc1:
`Input.get_joy_guid(0)` -> `__XINPUT_DEVICE__`

master (0a7942f4bb61cba8f02f2cbe0ec83b95eefa6726):
`Input.get_joy_guid(0)` -> `__XINPUT_DEVICE__`

**Comment by madmiraal** [comment]
On Windows, all XBox controllers should be picked up as XInput devices, irrespective of it's GUID. If Godot detects an XInput device, it just assigns it the XInput mapping. Godot doesn't even try to determine the GUID, which is why there is no GUID. So, aside from not returning the GUID when requested, the original issue is not an issue.

What is odd is the mapping used by Godot is so different from the mapping in the SDL database:
Godot:
```
__XINPUT_DEVICE__,XInput Gamepad,a:b12,b:b13,x:b14,y:b15,start:b4,back:b5,leftstick:b6,rightstick:b7,leftshoulder:b8,rightshoulder:b9,dpup:b0,dpdown:b1,dpleft:b2,dpright:b3,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:a4,righttrigger:a5,platform:Windows,
```

SDL database:
```
xinput,XInput Controller,a:b0,b:b1,back:b6,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,guide:b10,leftshoulder:b4,leftstick:b8,lefttrigger:a2,leftx:a0,lefty:a1,rightshoulder:b5,rightstick:b9,righttrigger:a5,rightx:a3,righty:a4,start:b7,x:b2,y:b3,platform:Windows,
```

However, as long as someone (@hoontee or @tristongilley?) can confirm that the mapping for the XBox controller works using the [Joypads Demo](https://github.com/godotengine/godot-demo-projects/tree/master/misc/joypads), I would suggest closing this issue.

Note: If there is still an issue in 3.2.3.rc1 (i.e. after #39060), with other, non-XBox controllers like the PS4 Controller, a separate issue should be opened. This issue is specific to XBox controllers.

**Comment by hoontee** [comment]
> However, as long as someone (hoontee or tristongilley?) can confirm that the mapping for the XBox controller works using the Joypads Demo, I would suggest closing this issue.

Yes, the gamepad works as expected on the `3.2` and `master` branches.

**Comment by madmiraal** [comment]
Closing because, on Windows all XBox One controllers should be detected as an XInput device i.e. this is working as expected.
